lib/booking.rb

```diff
@@ -0,0 +1,242 @@
+        end # .each
+      end # .each
+      return date_blocks
+    end # check_date_for_block
+
+    def return_all_block_ids
+      all_block_id = []
+      @all_blocks.each do |block|
+        all_block_id << block.block_id
+      end # .each
+     return all_block_id
+   end # return_all_block_ids
+
+    def check_block_availibility(block_id)
+      # will iterate though the @all_block array to find the block with the right ID
+      # if the block ID doesn't exsist
+      # if the ID does exist then it will rerurn the Block.block_rooms array
+      # if this array is empty then there is not avaibility
+      # if there are rooms in the array then there is still availible in the block
+      # Then this array can be passed to the reserve_from_block method so that you can book directly from the block
+      #Will have to have another method that will return true or false depending on the return array of the check_block_availibility method to be a simple way to check if there is availibility in the block
+      id = block_id.upcase
+      all_block_id = return_all_block_ids
+      if !(all_block_id.include?(id))
+        raise BookingError.new("That booking ID does not exist!")
+      end # if
+      block_rooms_remaining = []
+      @all_blocks.each do |block|
+        if block.block_id == id
+          block_rooms_remaining = block.block_rooms
+        end # if
+      end # .each
+      return block_rooms_remaining
+    end # def check_block_availibility
+
+    def rooms_left_in_block(block_id)
+      id = block_id.upcase
+      all_block_id = return_all_block_ids
+      if !(all_block_id.include?(id))
+        raise BookingError.new("That booking ID does not exist!")
+      end # if
+      rooms_left = check_block_availibility(block_id)
+      if rooms_left.length == 0
+        return "There are no rooms left in the #{id} block"
+      elsif rooms_left.length > 0
+        return "There are #{rooms_left.length} rooms left in the #{id} block"
+      end # if/elsif
+    end # rooms_left_in_block
+
+    def return_one_block(block_id)
+      # will return the block that has the ID that you're searching for
+      # this method will be called in the reserve_from_block method to reserve a room from a specific block!
+      id = block_id.upcase
+      all_block_id = return_all_block_ids
+      if !(all_block_id.include?(id))
+        raise BookingError.new("That booking ID does not exist!")
+      end # if
+      @all_blocks.each do |block|
+        if block.block_id == id
+          return block
+        end # if
+      end # .each
+    end # return_one_block
+
+    def reserve_from_block(block_id, num_of_rooms)
+      id = block_id.upcase
+      all_block_id = return_all_block_ids
+      if !(all_block_id.include?(id))
+        raise BookingError.new("That booking ID does not exist!")
+      end # if
+      reservation_id = @all_reservations.length + 1
+      block_reservations = return_one_block(id).reserve_room(reservation_id, num_of_rooms)
```

Comment:
I love that you call a method here rather than manipulating the block's data directly! Good example of loose coupling!